sbi/inference/posteriors/sbi_posterior.py

```diff
@@ -250,70 +235,108 @@ def _sample_posterior_mcmc(
             torch.Tensor of shape [num_samples, parameter_dim]
         """
 
-        # HMC and NUTS from Pyro.
-        # Defining the potential function as an object means Pyro's MCMC scheme
-        # can pickle it to be used across multiple chains in parallel, even if
-        # the potential function requires evaluating a neural likelihood as is the
-        # case here.
-        potential_function = NeuralPotentialFunction(
-            self, self.prior, context  # TODO: passing self not nice
-        )
+        if mcmc_method == "slice-np":
+            samples = self.slice_np_mcmc(num_samples, context, thin)
+        else:
+            samples = self.pyro_mcmc(num_samples, context, mcmc_method, thin)
 
-        # Axis-aligned slice sampling implementation in NumPy
-        target_log_prob = (
-            lambda parameters: self.log_prob(
-                inputs=torch.Tensor(parameters).reshape(1, -1),
-                context=context.reshape(1, -1),
+        # Back to training mode.
+        self.neural_net.train()
+
+        return samples
+
+    def slice_np_mcmc(self, num_samples, context, thin):
```

Comment:
The potential function for slice sampling with numpy is defined within the class, while the potential functions for Pyro as defined within the inference classes. Might be cleaner to move the potential functions here into proper functions (instead of lambdas) and define them inside the inference classes as well -- and pick them up here, as done further below